File: rust/src/equipment/tools/poisons.rs
use crate::core_mechanics::abilities::AbilityTag;
use crate::equipment::poison::poison_description;
use crate::equipment::poison::Exposure::{Contact, Ingestion, Injury};
use crate::equipment::poison::Form::{Gas, Liquid, Powder};
use crate::equipment::{Tool, ToolCategory};

fn poison() -> Tool {
    return Tool {
        category: ToolCategory::Poison,
        tags: vec![AbilityTag::Poison],
        ..Default::default()
    };
}

// The baseline for a consumable item of rank X is a spell of rank X+2.
//
// Normally, a "this round and next round" spell at Medium range is drX-2.
// "Every poison stage" is worse than "this round and next round" because it isn't
// guaranteed to deal damage next round, but it's better because it can deal damage three times
// total instead of twice, so call it equivalent.
// Since poison deals flat damage, that translates to drX-1.
// Apply -1dr if the poison is not removed at the second escalation.
//
// Damage that only applies against injured targets is weak as a concept, and will rarely reach a
// very high escalation count, so it is not removed on second escalation like normal poison damage.
// Applying this as a separate modifier instead of increasing the effective rank of injury poisons
// means that injury-only debuffs don't get too strong.
//
// Pure debuff "debuff while poisoned" poisons are +1 rank over a regular spell.
// Although they can be removed eventually, that's generally less important than getting three
// tries to succeed from a single usage, especially on elites.
// Poisons gain +1 rank if their stage 3 escalates the debuff.
//
// A powder poison typically requires a standard action to apply to an adjacent creature.
// Its effective spell rank is X+4 for contact/ingestion, never injury (including the +2 rank
// modifier for adjacent range).
// As a reminder, that means drX+1 is the standard damage value (dr5 -> dr3 from double hit -> dr2
// from flat)
//
// A liquid contact poison typically requires a non-action to apply with a weapon.
// Its effective spell rank is X for contact, X+2 for ingestion, X+4 for injury.
// As a reminder, that means drX-1 is the standard damage value, drX for ingestion, and drX+1 for
// injury.
//
// A gas poison typically requires a standard action to apply to a Tiny radius zone within Short range.
// Its effective spell rank is X+1 for contact and ingestion (which can be blocked by holding
// breath, which is sometimes beneficial to make them asymmetric).
//
// Applying a debuff only on the third escalation is not quite as big of a penalty as injured-only,
// but it's significant. Injured is 0.4x, say that third escalation is x0.67 EA.
// It could possibly be lower, but this gives room to include damage so the poison does something
// before the third escalation debuff.
// So stunned is 2 EA or r4.
pub fn poisons() -> Vec<Tool> {
    let mut poisons = vec![];

    // Contact powder means this is effective spell rank 4
    poisons.push(Tool {
        name: "Poison, Snakeroot".to_string(),
        rank: 0,
        short_description: "Repeatedly deals $dr1l damage".to_string(),
        description: poison_description(
            Contact,
            Powder,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr1l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Contact powder means this is effective spell rank 6, or 5 for a pure debuff.
    poisons.push(Tool {
        name: "Poison, Nightshade".to_string(),
        rank: 2,
        short_description: "Repeatedly deals $dr3l damage".to_string(),
        description: poison_description(
            Ingestion,
            Powder,
            r"
                The poison's accuracy is $consumableaccuracy-3.
                It inflicts $dr4l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Contact powder means this is effective spell rank 5
    // +4a for -1 flat dr
    poisons.push(Tool {
        name: "Poison, Wolfsbane".to_string(),
        rank: 1,
        short_description: "Repeatedly deals $dr1l damage".to_string(),
        description: poison_description(
            Contact,
            Powder,
            r"
                The poison's accuracy is $consumableaccuracy+4.
                It inflicts $dr1l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Contact liquid means this is effective spell rank 2.
    poisons.push(Tool {
        name: "Poison, Jellyfish Extract".to_string(),
        rank: 1,
        short_description: "Repeatedly deals $dr0l damage".to_string(),
        description: poison_description(
            Contact,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr0l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Ingestion liquid is drX
    poisons.push(Tool {
        name: "Poison, Baneberry".to_string(),
        rank: 1,
        short_description: "Repeatedly deals $dr1l damage".to_string(),
        description: poison_description(
            Ingestion,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr1l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // -3a for +1dr on flat damage
    poisons.push(Tool {
        name: "Poison, Tree Frog Coating".to_string(),
        rank: 2,
        short_description: "Repeatedly deals $dr2l damage".to_string(),
        description: poison_description(
            Contact,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy-3.
                It inflicts $dr2l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Bloodroot".to_string(),
        rank: 3,
        short_description: "Repeatedly deals $dr3l damage".to_string(),
        description: poison_description(
            Contact,
            Powder,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr4l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Ingestion powder means this is effective spell rank 6
    poisons.push(Tool {
        name: "Poison, Arsenic".to_string(),
        rank: 2,
        short_description: "Repeatedly deals $dr4l damage".to_string(),
        description: poison_description(
            Ingestion,
            Powder,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr3l damage immediately and with each escalation.
                The second escalation also ends the poison.
            ",
        ),
        ..poison()
    });

    // Contact liquid means this is effective spell rank 5
    poisons.push(Tool {
        name: "Poison, Dragon Bile".to_string(),
        rank: 4,
        short_description: "Endlessly deals $dr3l damage".to_string(),
        description: poison_description(
            Contact,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr3l damage immediately and with each escalation.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Mind Fog".to_string(),
        rank: 4,
        short_description: "Repeatedly deals $dr2l damage and eventually stuns".to_string(),
        description: poison_description(
            Ingestion,
            Gas,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr2l damage immediately and with each escalation.
                The second escalation also makes the target \stunned.
            ",
        ),
        ..poison()
    });

    // -1dr for +1a, endless damage
    poisons.push(Tool {
        name: "Poison, Black Lotus".to_string(),
        rank: 6,
        short_description: "Endlessly deals $dr5l damage".to_string(),
        description: poison_description(
            Contact,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy+1.
                It inflicts $dr5l damage per \glossterm<poison stage>.
            ",
        ),
        ..poison()
    });

    poisons.append(&mut injury_poisons());

    return poisons;
}

// These are stored in a separate function since they have different scaling.
// The baseline damage over time for a liquid injury poison would be drX+3.
// That's a bit absurd, so generally add +1 accuracy or make the poison endless.
fn injury_poisons() -> Vec<Tool> {
    let mut poisons = vec![];
    
    poisons.push(Tool {
        name: "Poison, Asp Venom".to_string(),
        rank: 1,
        short_description: "Stuns".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                A poisoned creature is \stunned while the poison lasts.
                The second escalation also inflicts $dr2l damage.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Giant Wasp Venom".to_string(),
        rank: 2,
        short_description: "Slows".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy+2.
                A poisoned creature is \slowed while the poison lasts.
                The second escalation also inflicts $dr3l damage.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Black Adder Venom".to_string(),
        rank: 2,
        short_description: "Endlessly deals $dr3l damage".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr3l damage immediately and with each escalation.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Wyvern Venom".to_string(),
        rank: 3,
        short_description: "Endlessly deals $dr3l damage".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy+4.
                It inflicts $dr3l damage immediately and with each escalation.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Blood Leech Venom".to_string(),
        rank: 4,
        short_description: "Endlessly deals $dr5l damage".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr5l damage immediately and with each escalation.
            ",
        ),
        ..poison()
    });

    poisons.push(Tool {
        name: "Poison, Purple Worm Venom".to_string(),
        rank: 5,
        short_description: "Endlessly deals $dr6l damage".to_string(),
        description: poison_description(
            Injury,
            Liquid,
            r"
                The poison's accuracy is $consumableaccuracy.
                It inflicts $dr6l damage immediately and with each escalation.
            ",
        ),
        ..poison()
    });

    return poisons;
}

File: typescript/src/equipment/data/consumables/poisons.ts
import { Tool, RawConsumable } from '../../types';
import { getPoisonDescription } from '../../poison';
import { parseDamageRank, parseAccuracyModifier } from '@src/abilities/spell_profile';

const DEBUFF_PATTERN =
  /\\(sickened|slowed|dazed|blinded|confused|dazzled|weakened|vulnerable|exposed|dread|frightened|panicked|immobilized|charmed|deafened|unsteady|frozen|helpless)|vital wound/i;

export function validatePoison(data: RawConsumable): string[] {
  const warnings: string[] = [];

  // Debuff parsing is out of scope for now
  if (DEBUFF_PATTERN.test(data.description)) {
    return warnings;
  }

  // Only validate damaging poisons with recurring damage
  const actualDr = parseDamageRank(data.description);
  if (actualDr === null || !data.description.includes('immediately and with each escalation')) {
    return warnings;
  }

  const deliveryMatch = data.description.match(
    /(contact|ingestion|injury)-based\s+(gas|liquid|pellet|powder)\s+poison/i,
  );
  if (!deliveryMatch) {
    warnings.push(`Poison "${data.name}": unable to parse delivery method (exposure and form).`);
    for (const warning of warnings) {
      console.warn(warning);
    }
    return warnings;
  }

  const exposure = deliveryMatch[1].toLowerCase();
  const form = deliveryMatch[2].toLowerCase();

  const accuracy = parseAccuracyModifier(data.description);
  let accuracyDrDelta = 0;
  if (accuracy === 0) {
    accuracyDrDelta = 0;
  } else if (accuracy === -3) {
    accuracyDrDelta = 1;
  } else if (accuracy === 4) {
    accuracyDrDelta = -1;
  } else {
    warnings.push(
      `Poison "${data.name}": invalid accuracy modifier (${accuracy > 0 ? '+' : ''}${accuracy}). Damaging poisons should only use -3 accuracy (for +1dr), +4 accuracy (for -1dr), or 0.`,
    );
  }

  let expectedBaseDr: number | null = null;
  if (form === 'powder') {
    if (exposure === 'contact' || exposure === 'ingestion') {
      expectedBaseDr = data.rank + 1;
    }
  } else if (form === 'liquid') {
    if (exposure === 'contact') {
      expectedBaseDr = data.rank - 1;
    } else if (exposure === 'injury' || exposure === 'ingestion') {
      expectedBaseDr = data.rank + 1;
    }
  } else if (form === 'gas') {
    if (exposure === 'contact' || exposure === 'ingestion') {
      expectedBaseDr = data.rank - 1;
    }
  }

  if (expectedBaseDr === null) {
    warnings.push(
      `Poison "${data.name}": unhandled exposure/form combination (${exposure} ${form}).`,
    );
  } else {
    const expectedDr = expectedBaseDr + accuracyDrDelta;
    if (actualDr !== expectedDr) {
      warnings.push(
        `Poison "${data.name}" (Rank ${data.rank}, ${exposure} ${form}) deals $dr${actualDr}l damage, but expected $dr${expectedDr}l.`,
      );
    }
  }

  for (const warning of warnings) {
    console.warn(warning);
  }
  return warnings;
}

function createPoison(data: RawConsumable): Tool {
  validatePoison(data);
  return {
    category: 'Poison',
    item: {
      magical: false,
      rarity: 'Common',
      tags: data.tags || ['Poison'],
      upgrades: data.upgrades || [],
      ...data,
    },
  };
}

// The baseline for a consumable item of rank X is a spell of rank X+2.
//
// Normally, a "this turn and next turn" spell at Medium range is drX-2.
// "Every poison stage" is worse than "this turn and next turn" because it isn't
// guaranteed to deal damage next turn, but it's better because it can deal damage more often
// total, so call it equivalent. This is a bit generous, but eh.
// That translates to drX+2-2 = drX by default.
//
// Injury-only damage is normally +2 ranks. That translates to drX+1 by default.
// Injury-only poisons are worse than normal poisons because most of the power of a poison is its potential to last for a long time.
// Equivalently, injury-only damage is typically written as an instant execution effect, which doesn't fit for poison.
// Therefore, injury-only poisons get an extra +2dr, or +1dr flat, so drX+2 by default (ignoring application method).
//
// Pure debuff "debuff while poisoned" poisons are equivalent to a condition.
// The successes to resist a poison are similar enough to the automatic elite condition removal.
// Poisons cost +1 rank if their second escalation has a stronger debuff, since that's more likely to work
// than stronger debuffs on crits, which is the default.

// A powder poison typically requires a standard action to apply to an adjacent creature.
// Its effective spell rank is X+4 for contact/ingestion, never injury (including the +2 rank
// modifier for adjacent range).
// As a reminder, that means drX+1 is the standard damage value (+4 -> +2 from double hit -> +1
// from flat)
//
// A liquid contact poison typically requires a non-action to apply with a weapon.
// Apply a -2 effect rank penalty for the non-action combat application, and ignore range modifier.
// A liquid poison's effective spell rank is X for contact or injury, or X+2 for ingestion??
// Liquid ingestion is basically exclusive to non-combat, so give it +1dr.
// That means drX-1 is the standard damage value for contact, drX+1 for injury, and drX+1 for ingestion.
//
// A gas poison typically requires a standard action to apply to a Tiny radius zone within Short range.
// Its effective spell rank is X+1 for contact and ingestion (which can be blocked by holding
// breath, which is sometimes beneficial to make them asymmetric).
//
// Applying a debuff only on the second escalation is not quite as big of a penalty as injured-only,
// but it's significant. Injured is 0.4x, say that second escalation is x0.67 EA.
// It could possibly be lower, but this gives room to include damage so the poison does something
// before the second escalation debuff?
// So dazed is 2 EA or rank 4.
export function poisons(): Tool[] {
  return [
    createPoison({
      name: 'Poison, Snakeroot',
      rank: 0,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr1l damage',
      description: getPoisonDescription(
        'contact',
        'powder',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr1l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Nightshade',
      rank: 2,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr3l damage',
      description: getPoisonDescription(
        'ingestion',
        'powder',
        `
          The poison's accuracy is $consumableaccuracy-3.
          It inflicts $dr4l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Wolfsbane',
      rank: 1,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr1l damage',
      description: getPoisonDescription(
        'contact',
        'powder',
        `
          The poison's accuracy is $consumableaccuracy+4.
          It inflicts $dr1l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Jellyfish Extract',
      rank: 1,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr0l damage',
      description: getPoisonDescription(
        'contact',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr0l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Baneberry',
      rank: 1,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr2l damage',
      description: getPoisonDescription(
        'ingestion',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr2l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Tree Frog Coating',
      rank: 2,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr2l damage',
      description: getPoisonDescription(
        'contact',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy-3.
          It inflicts $dr2l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Bloodroot',
      rank: 3,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr3l damage',
      description: getPoisonDescription(
        'contact',
        'powder',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr4l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Arsenic',
      rank: 2,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr3l damage',
      description: getPoisonDescription(
        'ingestion',
        'powder',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr3l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Dragon Bile',
      rank: 4,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr3l damage',
      description: getPoisonDescription(
        'contact',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr3l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Mind Fog',
      rank: 4,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr2l damage and eventually stuns',
      description: getPoisonDescription(
        'ingestion',
        'gas',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr2l damage immediately and with each escalation.
          The second escalation also makes the target \\dazed.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Black Lotus',
      rank: 6,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr5l damage',
      description: getPoisonDescription(
        'contact',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr5l damage immediately and with each escalation.
        `,
      ),
    }),
    ...injuryPoisons(),
  ];
}

// These are stored in a separate function since they have different scaling.
// As noted above, the baseline damage over time for a liquid injury poison is
// drX+1 flat damage.
function injuryPoisons(): Tool[] {
  return [
    createPoison({
      name: 'Poison, Asp Venom',
      rank: 1,
      attunement: 'Unrestricted',
      short_description: 'Sickens',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          A poisoned creature is \\sickened while the poison lasts.
          The second escalation also inflicts $dr2l damage.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Formian Drone Venom',
      rank: 1,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr2l damage',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr2l damage immediately and with each escalation.
          The second escalation also ends the poison.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Giant Wasp Venom',
      rank: 2,
      attunement: 'Unrestricted',
      short_description: 'Slows',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy+2.
          A poisoned creature is \\slowed while the poison lasts.
          The second escalation also inflicts $dr3l damage.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Black Adder Venom',
      rank: 2,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr3l damage',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr3l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Wyvern Venom',
      rank: 3,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr4l damage',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr4l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Blood Leech Venom',
      rank: 4,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr4l damage',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy+4.
          It inflicts $dr4l damage immediately and with each escalation.
        `,
      ),
    }),
    createPoison({
      name: 'Poison, Purple Worm Venom',
      rank: 5,
      attunement: 'Unrestricted',
      short_description: 'Repeatedly deals $dr6l damage',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          It inflicts $dr6l damage immediately and with each escalation.
        `,
      ),
    }),
    // Unclear rank for the special effect
    createPoison({
      name: 'Poison, Frostweb Spider Venom',
      rank: 5,
      attunement: 'Unrestricted',
      short_description: 'Slows and freezes',
      description: getPoisonDescription(
        'injury',
        'liquid',
        `
          The poison's accuracy is $consumableaccuracy.
          A poisoned creature is \\slowed while the poison lasts.
          The second escalation also inflicts a \\glossterm{vital wound} with a unique vital wound effect.
          Instead of making a \\glossterm{vital roll} for the \\glossterm{vital wound}, the target becomes deathly cold.
          % TODO: is this measured from the target's turn or the spider's turn?
          Whenever it takes damage from a \\atCold ability, it becomes \\briefly \\helpless.
          This effect lasts until the vital wound is removed.
        `,
      ),
    }),
  ];
}
